cni-plugin/linkerd-cni-validation/src/main.rs

```diff
@@ -0,0 +1,170 @@
+                info!(%client_addr, "accepted connection");
+                tokio::spawn(async move {
+                    let bytes = REDIRECT_RESPONSE.as_bytes();
+                    match stream.write_all(&bytes).await {
+                        Ok(_) => debug!(%client_addr, "wrote {:?} bytes", &bytes.len()),
+                        Err(e) => error!(%client_addr, "failed to write bytes to client: {:?}", e),
+                    }
+                });
+            }
+        })
+        .instrument(tracing::info_span!("srv_handler"));
+
+        tokio::select! {
+            _shutdown = shutdown_rx.signaled() => {
+                return Ok(())
+            }
+            _v = handler => {
+                return Ok(())
+            }
+        };
+    })
+    .instrument(tracing::info_span!("validation_server"));
+
+    let client_handle = tokio::spawn(async move {
+        let timeout = std::time::Duration::from_secs(120);
+        if let Err(_) = tokio::time::timeout(timeout, ready_rx).await {
+            return Err::<(), anyhow::Error>(anyhow!(
+                "timed-out (120s) waiting for server to be ready"
+            ));
+        };
+
+        let target_addr = SocketAddr::new(IpAddr::V4(target_host), target_port);
+        let mut stream = {
+            debug!(original_dst = %target_addr, "building validation client");
+            let socket = TcpStream::connect(target_addr)
+                .await
+                .with_context(|| format!("failed to connect: {}", target_addr))?;
+            let peer = socket.peer_addr()?;
+            debug!(original_dst = %peer, "client connected to validation server");
+            socket
+        };
+
+        tokio::select! {
+            is_readable = stream.readable() => {
+                if let Err(e) = is_readable {
+                    anyhow::bail!("cannot read off client socket {}", e);
+                }
+            }
+
+            () = tokio::time::sleep(timeout) => {
+                anyhow::bail!("timed-out (120s) waiting for socket to become readable");
+            }
+        };
+
+        let mut buf = [0u8; 100];
+        let read_sz = stream.read(&mut buf[..REDIRECT_RESPONSE.len()]).await?;
+        let resp = String::from_utf8(buf[..REDIRECT_RESPONSE.len()].to_vec())?;
+        debug!(redirect_response=%resp, "read {} bytes", read_sz);
+
+        if resp == REDIRECT_RESPONSE {
+            return Ok(());
+        } else {
+            anyhow::bail!(
+                "expected client to receive {:?}, got {:?} instead",
+                REDIRECT_RESPONSE,
+                resp,
+            );
+        }
+    })
+    .instrument(tracing::info_span!("validation_client"));
+
+    let result = client_handle.await?;
+    match result {
+        Ok(_) => {
+            info!("passed validation: iptables rules are properly set-up");
+            shutdown_server(shutdown_tx).await;
+        }
+        Err(e) => {
+            error!("failed validation: {}", e);
```

Comment:
nit: consider recording the error as a structured field?
```suggestion
            error!(error = %e, "failed validation");
```